evennia/contrib/aws-s3-cdn.py

```diff
@@ -0,0 +1,847 @@
+                )
+            else:
+                raise
+
+    def close(self):
+        if self._is_dirty:
+            self._flush_write_buffer()
+            # TODO: Possibly cache the part ids as they're being uploaded
+            # instead of requesting parts from server. For now, emulating
+            # s3boto's behavior.
+            parts = [{'ETag': part.e_tag, 'PartNumber': part.part_number}
+                     for part in self._multipart.parts.all()]
+            self._multipart.complete(
+                MultipartUpload={'Parts': parts})
+        else:
+            if self._multipart is not None:
+                self._multipart.abort()
+            if 'w' in self._mode and self._raw_bytes_written == 0:
+                self._create_empty_on_close()
+        if self._file is not None:
+            self._file.close()
+            self._file = None
+
+
+@deconstructible
+class S3Boto3Storage(Storage):
+    """
+    Amazon Simple Storage Service using Boto3
+    This storage backend supports opening files in read or write
+    mode and supports streaming(buffering) data in chunks to S3
+    when writing.
+    """
+    default_content_type = 'application/octet-stream'
+    # If config provided in init, signature_version and addressing_style settings/args are ignored.
+    config = None
+
+    # used for looking up the access and secret key from env vars
+    access_key_names = ['AWS_S3_ACCESS_KEY_ID', 'AWS_ACCESS_KEY_ID']
+    secret_key_names = ['AWS_S3_SECRET_ACCESS_KEY', 'AWS_SECRET_ACCESS_KEY']
+    security_token_names = ['AWS_SESSION_TOKEN', 'AWS_SECURITY_TOKEN']
+    security_token = None
+
+    access_key = setting('AWS_S3_ACCESS_KEY_ID', setting('AWS_ACCESS_KEY_ID'))
+    secret_key = setting('AWS_S3_SECRET_ACCESS_KEY', setting('AWS_SECRET_ACCESS_KEY'))
+    file_overwrite = setting('AWS_S3_FILE_OVERWRITE', True)
+    object_parameters = setting('AWS_S3_OBJECT_PARAMETERS', {})
+    bucket_name = setting('AWS_STORAGE_BUCKET_NAME')
+    auto_create_bucket = setting('AWS_AUTO_CREATE_BUCKET', False)
+    default_acl = setting('AWS_DEFAULT_ACL', 'public-read')
+    bucket_acl = setting('AWS_BUCKET_ACL', default_acl)
+    querystring_auth = setting('AWS_QUERYSTRING_AUTH', True)
+    querystring_expire = setting('AWS_QUERYSTRING_EXPIRE', 3600)
+    signature_version = setting('AWS_S3_SIGNATURE_VERSION')
+    reduced_redundancy = setting('AWS_REDUCED_REDUNDANCY', False)
+    location = setting('AWS_LOCATION', '')
+    encryption = setting('AWS_S3_ENCRYPTION', False)
+    custom_domain = setting('AWS_S3_CUSTOM_DOMAIN')
+    addressing_style = setting('AWS_S3_ADDRESSING_STYLE')
+    secure_urls = setting('AWS_S3_SECURE_URLS', True)
+    file_name_charset = setting('AWS_S3_FILE_NAME_CHARSET', 'utf-8')
+    gzip = setting('AWS_IS_GZIPPED', False)
+    preload_metadata = setting('AWS_PRELOAD_METADATA', False)
+    gzip_content_types = setting('GZIP_CONTENT_TYPES', (
```

Comment:
This is an interesting point - people used to Evennia Attributes may think that they could store anything (including raw Python structures/objects) in this bucket in a similar way. It should be made clear the limitations of S3 in this regard.